### PR TITLE
Add usage of `EMParams`

### LINKED ISSUE #1026: Use `EMParams` in `ExpectationMaximizationModel`
## Description

The `ExpectationMaximizationModel` needs to be refactored to use the `EMParams` data class.

### PR DESCRIPTION
**Pull request by johschnee**
## Changes

- Class `EMParams` is now used by the `ExpectationMaximizationModel` in `tools/_em_model.py`.

## Related issues

Closes #1026, #1025.

### CODE REVIEW

**Comment by johschnee** [comment]
@WeilerP Currently, I have the following questions about the future design of the API:
1. Should all the fitted parameters still be written in the AnnData object by default when calling `ExpectationMaximiziationModel.fit()`? Or should the `EMParams` class become the new way of handling the inferred parameters and only be written to the AnnData object when calling `ExpectationMaximiziationModel.export_results_adata()` (not yet implemented)?
2. My current implementation relies on the methods `_read_pars()` and `_write_pars()` in `tools/_em_model_core.py`. Should these methods be kept in the implementation at all or do you aim to remove them with this restructuring?
3. In the old method `recover_dynamics()` the parameter copy is false by default. In the new implementation `ExpectationMaximizationModel.fit()` it is true by default. Is this on purpose?

File: scvelo/tools/_em_model.py [hunk 1]
```diff
@@ -126,6 +125,10 @@ def __init__(
         self._n_jobs = get_n_jobs(n_jobs=n_jobs)
         self._backend = backend
 
+    def _initialize_state_dict(self, adata, pars_names=None, key="fit"):
+        pars = _read_pars(adata, pars_names, key)
```
Comment (WeilerP) [inline scvelo/tools/_em_model.py:129]: Can you please refactor the code s.t. we do not call `_read_pars`? We'll remove `_em_model_core.py` and `_em_model_utils.py` in the future.

File: scvelo/tools/_em_model.py [hunk 2]
```diff
@@ -274,7 +264,7 @@ def fit(
             "use_raw": use_raw,
         }
 
-        _write_pars(adata, _pars, add_key=self._fit_key)
+        _write_pars(adata, list(asdict(sd).values()), add_key=self._fit_key)
```
Comment (WeilerP) [inline scvelo/tools/_em_model.py:267]: Same as for `_read_pars`. Also, let's use the `export_results_adata` function for writing parameters and not have it part of the `fit` method.

**Comment by johschnee** [comment]
Yes, I was working on it actually today. I added the implementation of the abstract methods to this PR, because `export_results_adata` is used instead of `_write_pars`.

File: scvelo/tools/_em_model.py (hunk 1 above)
Comment (johschnee) [reply scvelo/tools/_em_model.py]: Of course, this refactoring is done.

File: scvelo/tools/_em_model.py (hunk 2 above)
Comment (johschnee) [reply scvelo/tools/_em_model.py]: I used `export_results_adata`, but I noticed that `fit` does not work without writing the parameters during fitting (or at least it would be a larger refactoring, which I'd want to do in a separate commit). The parameters written to the adata object are used by the `align_dynamics` method, which is called during `fit` after writing.

`align_dynamics` also modifies the written values. Hence, `export_results_adata` currently gives different results than there are returned by `fit`.

For refactoring, I'd suggest to modify `align_dynamics` as well such that is uses `EMParams` and copy it to `_em_model.py` (currently the method is in `_em_model_core.py`). Is this approach okay or would you suggest something different?

File: scvelo/tools/_em_model.py [hunk 3]
```diff
@@ -158,15 +168,17 @@ def _prepare_genes(self):
             var_names = var_names[np.argsort(np.sum(X, 0))[::-1][: self._n_top_genes]]
         self._var_names = var_names
 
-    # TODO: Implement
     def state_dict(self):
         """Return the state of the model."""
-        raise NotImplementedError
+        return asdict(self._state_dict)
```
Comment (WeilerP) [inline scvelo/tools/_em_model.py:173]: Can you please undo this implementation? This PR should only concern switching from the current implementation to using `EMParams`. The implementation of these functions is covered in #1025.

File: scvelo/tools/_em_model.py [hunk 4]
```diff
@@ -126,6 +119,23 @@ def __init__(
         self._n_jobs = get_n_jobs(n_jobs=n_jobs)
         self._backend = backend
 
+    def _initialize_state_dict(
```
Comment (WeilerP) [inline scvelo/tools/_em_model.py:122]: Why do we need the argument `parameters` and `key` if we never pass them to the function?

File: scvelo/tools/_em_model.py [hunk 5]
```diff
@@ -158,15 +168,17 @@ def _prepare_genes(self):
...
     def state_dict(self):
         """Return the state of the model."""
-        raise NotImplementedError
+        return asdict(self._state_dict)
 
-    # TODO: Implement
-    def export_results_adata(self):
-        """Export the results to the AnnData object."""
-        raise NotImplementedError
+    def export_results_adata(self, copy: bool = True, add_key: str = "fit"):
```
Comment (WeilerP) [inline scvelo/tools/_em_model.py:175]: See my comment for the `state_dict` function: These functions should be implemented in a separate PR for maximal readability, etc.

File: scvelo/tools/_em_model.py [hunk 6]
```diff
@@ -239,42 +249,30 @@ def fit(
...
-        adata = self._adata.copy() if copy else self._adata
+        adata = self.export_results_adata(copy=copy, add_key=self._fit_key)
 
         adata.uns["recover_dynamics"] = {
             "fit_connected_states": fit_connected_states,
             "fit_basal_transcription": self._fit_basal_transcription,
             "use_raw": use_raw,
         }
 
-        _write_pars(adata, _pars, add_key=self._fit_key)
```
Comment (WeilerP) [inline scvelo/tools/_em_model.py:277]: Similar to the implemented classes, can you please revert these changes and open a separate PR for them? Or am I missing something, and all of it is needed when switching to using `EMParams`?

**Review by WeilerP — COMMENTED** [review]
@johschnee, I left another few comments. Mainly two things:
1. I updated/refactored the `_initialize_state_dict` function. Can you please double check that it is still working?
2. Can you please und all changes irrelevant to this PR (_i.e._, refactoring the code to rely on `EMParams`)? These changes (_e.g._, function implementations) should be there own PR accompanied by an issue.

**Comment by WeilerP** [comment]
FYI, @johschnee, the _ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (4,) + inhomogeneous part._ error is caused by a new Numpy version (see #1058).

**Comment by johschnee** [comment]
Hi @WeilerP,
this branch now contains the changes we discussed. In particular, I changed the following:
- Added metadata to the attributes of `EMParams` and read/write methods: The metadata allows a different handling of vector and matrix parameters without hardcoding them in the read/write methods. I decided to add the read/write methods to EMParams because they rely a lot on the attributes of EMParams.
- In the method `fit()`, I used the class EMParams. Since the AnnData object should not be modified during fit(), I defined some private attributes to store results and additional settings: `_loss`, `_fit_connected_states`, and `_use_raw`. They are written to an AnnData object when `export_results_adata()` is called.
- As we discussed, I copied `_align_dynamics()` to the class `ExpectationMaximizationModel` and adapted it slightly to use the EMParams. Now, `ExpectationMaximizationModel` only uses the class `DynamicsRecovery` from `_em_model_core`, but no other methods.

Overall, 
```
model = ExpectationMaximizationModel(adata)
model.fit()
adata_return = model.export_results_adata()
```
and
```
scv.tl.recover_dynamics(adata)
```
should lead to the same results. I tested it with simulated data and the pancreas dataset.

**Comment by johschnee** [comment]
Hi @WeilerP,
I changed the behavior of `export_to_adata()` such that arrays containing only nan values are not written anymore.
I also compared the results of the old and the new API for the simulated data and the pancreas dataset using numpy. As expected, they were identical.
Hence, it would be great if you could review this PR.

**Review by WeilerP — APPROVED** [review]
Thanks, @johschnee!
